### PR TITLE
[feature](load-refactor-with-tvf) S3 load with S3 tvf and native insert

### LINKED ISSUE #18476: [Feature] load code refactoring
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

# Motivation
Doris now supports many load methods. The implementation code of these methods is intertwined and the attributes overlap. To improve maintainability, we plan to refactor the code.

# Goals
long-term: unify all data loading methods (brokerload, insert, etc) into one
short-term:
1) unify the attribute specification method and parsing method and default values of various loads;
2) reuse single code base as much as possible and delete duplicate code;

# Plan Overview
We plan to use insert stmt as the common backend for all load methods. Various loads retain their original user interface, but will eventually be converted into insert stmt for planning and subsequent execution.

# Task Breakdown
Phase1: make S3/Broker load work on `insert stmt`

1) #18477
2) #18480
3) #18478

Phase2: make all other load method work on `insert stmt`

1) run streamload with insert stmt planner
2) support mysql load
3) support routine load
...

Phase3: fully support all use cases in all load method
1) insert stmt support writing multi-tables in single transaction
2) insert stmt support sequence column v1
...

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Hastyshell**
# Proposed changes

Issue Number: close #18476

## Problem summary

An implementation of S3 load based on S3 tvf and native insert, validate the feasibility. 

- [ ] column mapping
     - [x] column list
     - [x] set
     - [ ] column from path
- [x] clauses
     - [x] where
     - [x] preceding filter predicate
     - [x] partition
 - [ ] negative
 - [ ] merge type
      - [x] merge
      - [x] append
      - [ ] delete
 - [x] infomation
      - [x] data in file
      - [x] column terminated by
      - [x] format as
      - [x] json properties
 - [ ] load properties 
 

problems exists

| unsolved                                                   | remarks                                                      |
| :----------------------------------------------------------- | :----------------------------------------------------------- |
| load properties                                              | intend to be implemented in https://github.com/apache/doris/pull/20411 |
| column from path                                             | need to modify the behavior of tvfs, parse column from path in get schema stage, and add a `column from path`tvf param if needed |
| All the columns are in table but none of them is in file     | could not be covered with insert statement                     |
| delete                                                       | could not be covered with insert statement                     |
| negative                                                     | could not be covered with insert statement                     |

## Checklist(Required)

* [x] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.36 seconds
 stream load tsv:          429 seconds loaded 74807831229 Bytes, about 166 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.3 seconds inserted 10000000 Rows, about 131K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230522114858_clickbench_pr_148401.html

File: fe/fe-core/src/main/java/org/apache/doris/analysis/S3LoadStmt.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                Lists.newArrayList(orderByElement), LimitElement.NO_LIMIT
+        );
+        return new InsertSource(selectStmt);
+    }
+
+    private static TableRef buildTvfRef(DataDescription dataDescription, BrokerDesc brokerDesc) throws DdlException {
+        final Map<String, String> params = Maps.newHashMap();
+
+        final List<String> filePaths = dataDescription.getFilePaths();
+        Preconditions.checkState(filePaths.size() == 1, "there should be one and only file path");
```
Comment (carlvinhust2012) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/S3LoadStmt.java:78]:
```suggestion
        Preconditions.checkState(filePaths.size() == 1, "there should be one and only file path");
```Preconditions.checkState(filePaths.size() == 1, "there should be only one file path");

Comment (carlvinhust2012) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/S3LoadStmt.java:78]:
```suggestion
        Preconditions.checkState(filePaths.size() == 1, "there should be only one file path");
```

**Comment by carlvinhust2012** [comment]
how about add some test case ？

**Comment by TangSiyang2001** [comment]
> how about add some test case ？

S3 Load cases is located in load_p2, and manually test has been carried out yet.

```
mysql> LOAD LABEL test.orc_s33 (
    ->     DATA INFILE("s3://test-for-student-1308700295/regression/segcompaction/segcompaction.orc")
    ->     INTO TABLE orc_s3
    ->     FORMAT AS "ORC"
    ->     (col_0, col_1, col_2, col_3, col_4, col_5, col_6, col_7, col_8, col_9, col_10, col_11, col_12, col_13, col_14, col_15, col_16, col_17, col_18, col_19, col_20, col_21, col_22, col_23, col_24, col_25, col_26, col_27, col_28, col_29, col_30, col_31, col_32, col_33, col_34, col_35, col_36, col_37, col_38, col_39, col_40, col_41, col_42, col_43, col_44, col_45, col_46, col_47, col_48, col_49)
    -> ) WITH S3 (
    ->     "AWS_ENDPOINT" = "cos.ap-beijing.myqcloud.com",
    ->     "AWS_ACCESS_KEY" = "AKIDd9RVMzIOI0V7Wlnbr9JG0WrhJk28zc2H",
    ->     "AWS_SECRET_KEY"="4uWxMhqnW3Plz97sPjqlSUXO1RhokRuO",
    ->     "AWS_REGION" = "ap-beijing"
    -> );
Query OK, 5000000 rows affected (4 min 59.51 sec)
{'label':'orc_s33', 'status':'VISIBLE', 'txnId':'2004'}

mysql> select count(*) from orc_s3;
+----------+
| count(*) |
+----------+
| 10000000 |
+----------+

mysql> select * from orc_s3 limit 1\G;
*************************** 1. row ***************************
 col_0: 1
 col_1: Lemon
 col_2: Raspberry
 col_3: Lychee
 col_4: Avocado
 col_5: Mango
 col_6: Watermelon
 col_7: Watermelon
 col_8: Pineapple
 col_9: Banana
col_10: Pineapple
col_11: Banana
col_12: Kiwi
col_13: Lychee
col_14: Lemon
col_15: Lemon
col_16: Avocado
col_17: Watermelon
col_18: Banana
col_19: Blueberry
col_20: Grapes
col_21: Apple
col_22: Plum
col_23: Blueberry
col_24: Grapes
col_25: Lemon
col_26: Lemon
col_27: Strawberry
col_28: Strawberry
col_29: Strawberry
col_30: Cherry
col_31: Plum
col_32: Pineapple
col_33: Plum
col_34: Orange
col_35: Orange
col_36: Lemon
col_37: Lemon
col_38: Raspberry
col_39: Pineapple
col_40: Strawberry
col_41: Strawberry
col_42: Lemon
col_43: Cherry
col_44: Plum
col_45: Apple
col_46: Cherry
col_47: Avocado
col_48: Pineapple
col_49: Watermelon
1 row in set (1.49 sec)
```

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.65 seconds
 stream load tsv:          429 seconds loaded 74807831229 Bytes, about 166 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          76.0 seconds inserted 10000000 Rows, about 131K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230522154048_clickbench_pr_148534.html

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.65 seconds
 stream load tsv:          422 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.4 seconds inserted 10000000 Rows, about 130K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523025143_clickbench_pr_148601.html

**Comment by freemandealer** [comment]
really nice job! TVF is more powerful than ever in Doris.

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.21 seconds
 stream load tsv:          455 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          67.4 seconds inserted 10000000 Rows, about 148K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230621183735_clickbench_pr_166420.html

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.87 seconds
 stream load tsv:          468 seconds loaded 74807831229 Bytes, about 152 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          70.3 seconds inserted 10000000 Rows, about 142K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230622021606_clickbench_pr_166475.html

**Comment by TangSiyang2001** [comment]
run p1

**Review by dataroaring — APPROVED** [review]
LGTM

File: fe/fe-core/pom.xml [hunk 2]
```diff
@@ -482,12 +482,12 @@ under the License.
...
         </dependency>
         <dependency>
-            <groupId>org.apache.doris</groupId>
-            <artifactId>hive-catalog-shade</artifactId>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
         </dependency>
         <dependency>
-           <groupId>org.apache.httpcomponents</groupId>
-           <artifactId>httpclient</artifactId>
```
Comment (CalvinKirs) [inline fe/fe-core/pom.xml:490]: @CalvinKirs  The `doris-shade` release should contain this

Comment (TangSiyang2001) [reply fe/fe-core/pom.xml:490]: So I just make it consistent with master, right?

**Review by CalvinKirs — APPROVED** [review]

File: fe/fe-core/pom.xml (hunk 2 above)
Comment (CalvinKirs) [reply fe/fe-core/pom.xml:490]: I just remind myself:)It can also be done in[doris-shade](https://github.com/apache/doris-shade)if you are interested.

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**